### PR TITLE
kpatch-build: ignore .discard sections

### LINKED ISSUE #711: kpatch-build: changed section .rela.discard.unreachable not selected for inclusion
[example.patch.txt](https://github.com/dynup/kpatch/files/1038682/example.patch.txt)

Building the attached patch file against Jiri's [for-4.12/klp-hybrid-consistency-model](https://help.github.com/articles/searching-issues/) branch with RHEL7.3 tools, encountered the following kpatch-build error:

```
kpatch-build -s /root/linux --skip-cleanup example.patch
Skipping cleanup
Using source directory at /root/linux
Testing patch file
checking file fs/proc/array.c
checking file kernel/exit.c
checking file kernel/fork.c
Reading special section data
Building original kernel
Building patched kernel
Extracting new and modified ELF sections
fork.o: changed function: _do_fork
ERROR: exit.o: 1 unsupported section change(s)
exit.o: changed function: do_exit
exit.o: changed section .rela.discard.unreachable not selected for inclusion
/usr/local/libexec/kpatch/create-diff-object: unreconcilable difference
ERROR: array.o: 1 unsupported section change(s)
array.o: changed function: proc_pid_status
array.o: changed section .rela.discard.unreachable not selected for inclusion
/usr/local/libexec/kpatch/create-diff-object: unreconcilable difference
ERROR: 2 error(s) encountered. Check /root/.kpatch/build.log for more details.
```

### PR DESCRIPTION
**Pull request by joe-lawrence**
Fixes #711.

Signed-off-by: Joe Lawrence <joe.lawrence@redhat.com>

### CODE REVIEW

**Comment by joe-lawrence** [comment]
Tested with this small patch:
```
diff --git a/fs/proc/array.c b/fs/proc/array.c
index 88c355574aa0..bc7219ed08ee 100644
--- a/fs/proc/array.c
+++ b/fs/proc/array.c
@@ -356,6 +356,7 @@ static inline void task_seccomp(struct seq_file *m, struct task_struct *p)
 static inline void task_context_switch_counts(struct seq_file *m,
                                                struct task_struct *p)
 {
+       seq_printf(m, "kpatch!\n");
        seq_put_decimal_ull(m, "voluntary_ctxt_switches:\t", p->nvcsw);
        seq_put_decimal_ull(m, "\nnonvoluntary_ctxt_switches:\t", p->nivcsw);
        seq_putc(m, '\n');
```
Before the change, kpatch-build would fail like so:
```
ERROR: array.o: 1 unsupported section change(s)
array.o: changed function: proc_pid_status
array.o: changed section .rela.discard.unreachable not selected for inclusion
/usr/local/libexec/kpatch/create-diff-object: unreconcilable difference
```

File: kpatch-build/kpatch-elf.c [hunk 1]
```diff
@@ -829,6 +829,8 @@ void kpatch_elf_teardown(struct kpatch_elf *kelf)
 	struct rela *rela, *saferela;
 
 	list_for_each_entry_safe(sec, safesec, &kelf->sections, list) {
+		if (sec->ignore)
+			continue;
```
Comment (jpoimboe) [inline kpatch-build/kpatch-elf.c]: Why was this part needed?

Comment (joe-lawrence) [reply kpatch-build/kpatch-elf.c]: Without it...

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff78683dd in __isoc99_fscanf () from /lib64/libc.so.6
Missing separate debuginfos, use: dnf debuginfo-install elfutils-libelf-0.168-1.fc24.x86_64 zlib-1.2.8-10.fc24.x86_64
(gdb) bt
#0  0x00007ffff78683dd in __isoc99_fscanf () from /lib64/libc.so.6
#1  0x000000000040d2e6 in symvers_read (table=0x733d60, path=0x7fffffffe73d "/root/.kpatch/obj/Module.symvers")
    at lookup.c:277
#2  0x000000000040d553 in lookup_open (obj_path=0x7fffffffe70c "/root/.kpatch/tmp/vmlinux", 
    symvers_path=0x7fffffffe73d "/root/.kpatch/obj/Module.symvers", hint=0x7ffff7fc0589 "array.c", locals=0x733ca0)
    at lookup.c:322
#3  0x0000000000408e53 in main (argc=7, argv=0x7fffffffe3f8) at create-diff-object.c:2738
```

Give me some time to backtrace to figure out why it's only needed after the patch.  I imagine there is something special about the discard sections vs. the previous use of the `sec->ignore` flag.

Comment (joe-lawrence) [reply kpatch-build/kpatch-elf.c]: Without it:
```
*** Error in `/root/kpatch.joe/kpatch-build/create-diff-object': free(): invalid next size (fast): 0x00000000006b5e90 ***
...
Program received signal SIGABRT, Aborted.
0x00007ffff7833738 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: dnf debuginfo-install libgcc-6.3.1-1.fc24.x86_64
(gdb) bt
#0  0x00007ffff7833738 in raise () from /lib64/libc.so.6
#1  0x00007ffff783534a in abort () from /lib64/libc.so.6
#2  0x00007ffff78746f0 in __libc_message () from /lib64/libc.so.6
#3  0x00007ffff787ce7a in _int_free () from /lib64/libc.so.6
#4  0x00007ffff78803dc in free () from /lib64/libc.so.6
#5  0x000000000040c659 in kpatch_elf_teardown (kelf=0x6a9630) at kpatch-elf.c:918
#6  0x0000000000408ffd in main (argc=7, argv=0x7fffffffe408) at create-diff-object.c:2794
```
Give me some time to backtrack to figure out why this patch's use of `sec->ignore` is different than before.

**Comment by joe-lawrence** [comment]
@jpoimboe , I'll need a little ELF help with this one.  That modification to `kpatch_elf_teardown` you asked about is (I think) papering over a problem in `kpatch_mark_ignored_sections_same`.  Since I was marking sections like `.rela.discard.unreachable` as ignored, I think the code in `kpatch_mark_ignored_sections_same` was clobbering over the section union for rela cases.  If I add in a check for `is_rela_section()`, I don't see the kfree problem.

**Comment by jpoimboe** [comment]
👍

**Pull request merged**